File: Services/FontManager.vb
Imports System.Diagnostics
Imports System.Globalization
Imports System.IO
Imports System.Windows
Imports System.Windows.Media

Namespace Services
    Public NotInheritable Class FontManager
        Private Const SpaceGroteskDirectory As String = "founts/SpaceGrotesk/ttf/static/"
        Private Const SpaceGroteskFamilyName As String = "Space Grotesk"
        Private Shared ReadOnly ApplicationFontBaseUri As New Uri("pack://application:,,,/")

        Private Shared ReadOnly _instance As New FontManager()
        Private Shared ReadOnly _defaultUiFallback As New FontFamily("Segoe UI")
        Private Shared ReadOnly _defaultChineseFallback As New FontFamily("Microsoft YaHei UI")

        Private _initialized As Boolean

        Private Sub New()
        End Sub

        Public Shared ReadOnly Property Instance As FontManager
            Get
                Return _instance
            End Get
        End Property

        Public Sub Initialize()
            If _initialized Then
                Return
            End If

            _initialized = True
            AddHandler LocalizationManager.Instance.LanguageChanged, AddressOf OnLanguageChanged
            ApplyCurrentLanguageFonts()
        End Sub

        Private Sub OnLanguageChanged(sender As Object, e As EventArgs)
            ApplyCurrentLanguageFonts()
        End Sub

        Private Sub ApplyCurrentLanguageFonts()
            If Application.Current Is Nothing Then
                Return
            End If

            Dim cultureName = LocalizationManager.Instance.CurrentCulture.Name
            Dim isChinese = cultureName.StartsWith("zh", StringComparison.OrdinalIgnoreCase)
            Dim displaySansBase = ResolveSpaceGrotesk()
            Dim chineseUi = ResolveChineseUiFont()
            Dim displaySans = ComposeFontFamily(displaySansBase, displaySansBase, chineseUi, _defaultUiFallback)
            Dim listDisplay = displaySans
            Dim body = If(isChinese, chineseUi, displaySans)
            Dim meta = body
            Dim microUi = If(isChinese, ResolveSmallChineseUiFont(), displaySans)

            Application.Current.Resources("Font.Meta") = meta
            Application.Current.Resources("Font.MicroUi") = microUi
            Application.Current.Resources("Font.Body") = body
            Application.Current.Resources("Font.DisplaySans") = displaySans
            Application.Current.Resources("Font.ListDisplay") = listDisplay
            Application.Current.Resources("Font.Serif") = body
            Application.Current.Resources("Font.Mono") = displaySans
            Application.Current.Resources("Font.EditorialSerif") = body
            Application.Current.Resources("Font.EditorialDisplay") = displaySans

            If ShouldWriteDiagnostics() Then
                WriteDiagnostics(displaySans, listDisplay, body, meta, microUi)
            End If
        End Sub

        Private Shared Function ResolveSpaceGrotesk() As FontFamily
            Return ResolveBundledFontFamily(SpaceGroteskDirectory,
                                            SpaceGroteskFamilyName,
                                            _defaultUiFallback)
        End Function

        Private Shared Function ResolveChineseUiFont() As FontFamily
            Return _defaultChineseFallback
        End Function

        Private Shared Function ResolveSmallChineseUiFont() As FontFamily
            Return _defaultChineseFallback
        End Function

        Private Shared Function ResolveBundledFontFamily(relativeDirectory As String,
                                                         familyName As String,
                                                         fallback As FontFamily) As FontFamily
            Dim normalizedDirectory = NormalizeFontDirectory(relativeDirectory)
            If Not HasBundledFontResources(normalizedDirectory) Then
                Trace.WriteLine(String.Format(CultureInfo.InvariantCulture,
                                              "FontManager fallback: no bundled font assets found under {0}",
                                              normalizedDirectory))
                Return fallback
            End If

            Dim relativeReference = String.Format(CultureInfo.InvariantCulture,
                                                  "./{0}#{1}",
                                                  normalizedDirectory,
                                                  familyName)

            Return CreateFontFamily(New String() {relativeReference}, fallback)
        End Function

        Private Shared Function ComposeFontFamily(fallback As FontFamily,
                                                  ParamArray families As FontFamily()) As FontFamily
            If families Is Nothing OrElse families.Length = 0 Then
                Return fallback
            End If

            Dim sources =
                families.Where(Function(item) item IsNot Nothing AndAlso Not String.IsNullOrWhiteSpace(item.Source)).
                         Select(Function(item) item.Source).
                         Distinct(StringComparer.OrdinalIgnoreCase)

            Return CreateFontFamily(sources, fallback)
        End Function

        Private Shared Function CreateFontFamily(sources As IEnumerable(Of String), fallback As FontFamily) As FontFamily
            If sources Is Nothing Then
                Return fallback
            End If

            Dim normalizedSources =
                sources.Where(Function(item) Not String.IsNullOrWhiteSpace(item)).
                        Distinct(StringComparer.OrdinalIgnoreCase).
                        ToArray()

            If normalizedSources.Length = 0 Then
                Return fallback
            End If

            Dim reference = String.Join(", ", normalizedSources)

            Try
                Return New FontFamily(ApplicationFontBaseUri, reference)
            Catch ex As Exception
                Trace.WriteLine(String.Format(CultureInfo.InvariantCulture,
                                              "FontManager fallback: failed to compose font family {0}. {1}",
                                              reference,
                                              ex.Message))
                Return fallback
            End Try
        End Function

        Private Shared Function HasBundledFontResources(normalizedDirectory As String) As Boolean
            Dim directoryReference = "./" & normalizedDirectory

            Try
                Return Fonts.GetFontFamilies(ApplicationFontBaseUri, directoryReference).Any()
            Catch ex As Exception
                Trace.WriteLine(String.Format(CultureInfo.InvariantCulture,
                                              "FontManager fallback: failed to enumerate bundled fonts under {0}. {1}",
                                              normalizedDirectory,
                                              ex.Message))
                Return False
            End Try
        End Function

        Private Shared Function NormalizeFontDirectory(relativeDirectory As String) As String
            Dim normalized = relativeDirectory.Replace("\"c, "/"c).Trim("/"c)

            If Not normalized.EndsWith("/", StringComparison.Ordinal) Then
                normalized &= "/"
            End If

            Return normalized
        End Function

        Private Shared Function ShouldWriteDiagnostics() As Boolean
            Return String.Equals(Environment.GetEnvironmentVariable("CLIKSYNC_FONT_DIAGNOSTICS"),
                                 "1",
                                 StringComparison.Ordinal)
        End Function

        Private Shared Sub WriteDiagnostics(displaySans As FontFamily,
                                            listDisplay As FontFamily,
                                            body As FontFamily,
                                            meta As FontFamily,
                                            microUi As FontFamily)
            Try
                Dim lines As New List(Of String) From {
                    "Timestamp=" & DateTimeOffset.Now.ToString("O", CultureInfo.InvariantCulture),
                    "Culture=" & LocalizationManager.Instance.CurrentCulture.Name
                }

                AppendDiagnostics(lines, "Font.DisplaySans", displaySans)
                AppendDiagnostics(lines, "Font.ListDisplay", listDisplay)
                AppendDiagnostics(lines, "Font.Body", body)
                AppendDiagnostics(lines, "Font.Meta", meta)
                AppendDiagnostics(lines, "Font.MicroUi", microUi)

                Dim outputPath = Path.Combine(AppContext.BaseDirectory, "font-diagnostics.txt")
                File.WriteAllLines(outputPath, lines)
            Catch ex As Exception
                Trace.WriteLine("FontManager diagnostics failed: " & ex.Message)
            End Try
        End Sub

        Private Shared Sub AppendDiagnostics(lines As ICollection(Of String), key As String, family As FontFamily)
            If lines Is Nothing Then
                Return
            End If

            If family Is Nothing Then
                lines.Add(key & ".Source=<null>")
                Return
            End If

            lines.Add(key & ".Source=" & family.Source)

            Dim typeface As New Typeface(family, FontStyles.Normal, FontWeights.Normal, FontStretches.Normal)
            Dim glyph As GlyphTypeface = Nothing
            If typeface.TryGetGlyphTypeface(glyph) AndAlso glyph IsNot Nothing Then
                Dim glyphFamily = glyph.FamilyNames.FirstOrDefault(Function(entry) entry.Key.IetfLanguageTag = "en-US").Value
                If String.IsNullOrWhiteSpace(glyphFamily) Then
                    glyphFamily = glyph.FamilyNames.Values.FirstOrDefault()
                End If

                lines.Add(key & ".GlyphFamily=" & glyphFamily)
            Else
                lines.Add(key & ".GlyphFamily=<unresolved>")
            End If
        End Sub
    End Class
End Namespace

File: src/ClickSyncMouseTester/Services/FontManager.cs
#define TRACE
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Windows;
using System.Windows.Media;

namespace ClickSyncMouseTester.Services;

public sealed class FontManager
{
    private const string SpaceGroteskDirectory = "Assets/Fonts/Bundled/SpaceGrotesk/ttf/static/";

    private const string SpaceGroteskFamilyName = "Space Grotesk";

    private static readonly Uri ApplicationFontBaseUri = new Uri("pack://application:,,,/");

    private static readonly FontManager _instance = new FontManager();

    private static readonly FontFamily _defaultUiFallback = new FontFamily("Segoe UI");

    private static readonly FontFamily _defaultChineseFallback = new FontFamily("Microsoft YaHei UI");

    private bool _initialized;

    public static FontManager Instance => _instance;

    private FontManager()
    {
    }

    public void Initialize()
    {
        if (!_initialized)
        {
            _initialized = true;
            LocalizationManager.Instance.LanguageChanged += OnLanguageChanged;
            ApplyCurrentLanguageFonts();
        }
    }

    private void OnLanguageChanged(object sender, EventArgs e)
    {
        ApplyCurrentLanguageFonts();
    }

    private void ApplyCurrentLanguageFonts()
    {
        if (System.Windows.Application.Current == null)
        {
            return;
        }

        bool isChineseCulture = LocalizationManager.Instance.CurrentCulture.Name.StartsWith("zh", StringComparison.OrdinalIgnoreCase);
        FontFamily spaceGrotesk = ResolveSpaceGrotesk();
        FontFamily chineseUiFont = ResolveChineseUiFont();
        FontFamily latinDisplayFont = ComposeFontFamily(spaceGrotesk, spaceGrotesk, chineseUiFont, _defaultUiFallback);
        FontFamily listDisplayFont = latinDisplayFont;
        FontFamily bodyFont = isChineseCulture ? chineseUiFont : latinDisplayFont;
        FontFamily metadataFont = bodyFont;
        FontFamily microUiFont = isChineseCulture ? ResolveSmallChineseUiFont() : latinDisplayFont;

        System.Windows.Application.Current.Resources["Font.Meta"] = metadataFont;
        System.Windows.Application.Current.Resources["Font.MicroUi"] = microUiFont;
        System.Windows.Application.Current.Resources["Font.Body"] = bodyFont;
        System.Windows.Application.Current.Resources["Font.DisplaySans"] = latinDisplayFont;
        System.Windows.Application.Current.Resources["Font.ListDisplay"] = listDisplayFont;
        System.Windows.Application.Current.Resources["Font.Serif"] = bodyFont;
        System.Windows.Application.Current.Resources["Font.Mono"] = latinDisplayFont;
        System.Windows.Application.Current.Resources["Font.EditorialSerif"] = bodyFont;
        System.Windows.Application.Current.Resources["Font.EditorialDisplay"] = latinDisplayFont;
        if (ShouldWriteDiagnostics())
        {
            WriteDiagnostics(latinDisplayFont, listDisplayFont, bodyFont, metadataFont, microUiFont);
        }
    }

    private static FontFamily ResolveSpaceGrotesk()
    {
        return ResolveBundledFontFamily("Assets/Fonts/Bundled/SpaceGrotesk/ttf/static/", "Space Grotesk", _defaultUiFallback);
    }

    private static FontFamily ResolveChineseUiFont()
    {
        return _defaultChineseFallback;
    }

    private static FontFamily ResolveSmallChineseUiFont()
    {
        return _defaultChineseFallback;
    }

    private static FontFamily ResolveBundledFontFamily(string relativeDirectory, string familyName, FontFamily fallback)
    {
        string normalizedDirectory = NormalizeFontDirectory(relativeDirectory);
        if (!HasBundledFontResources(normalizedDirectory))
        {
            Trace.WriteLine(string.Format(CultureInfo.InvariantCulture, "FontManager fallback: no bundled font assets found under {0}", normalizedDirectory));
            return fallback;
        }

        string fontSource = string.Format(CultureInfo.InvariantCulture, "./{0}#{1}", normalizedDirectory, familyName);
        return CreateFontFamily(new[] { fontSource }, fallback);
    }

    private static FontFamily ComposeFontFamily(FontFamily fallback, params FontFamily[] families)
    {
        if (families == null || families.Length == 0)
        {
            return fallback;
        }
        return CreateFontFamily((from fontFamily in families
                                 where fontFamily != null && !string.IsNullOrWhiteSpace(fontFamily.Source)
                                 select fontFamily.Source).Distinct<string>(StringComparer.OrdinalIgnoreCase), fallback);
    }

    private static FontFamily CreateFontFamily(IEnumerable<string> sources, FontFamily fallback)
    {
        if (sources == null)
        {
            return fallback;
        }

        string[] uniqueSources = sources.Where(source => !string.IsNullOrWhiteSpace(source)).Distinct(StringComparer.OrdinalIgnoreCase).ToArray();
        if (uniqueSources.Length == 0)
        {
            return fallback;
        }

        string sourceList = string.Join(", ", uniqueSources);
        try
        {
            return new FontFamily(ApplicationFontBaseUri, sourceList);
        }
        catch (Exception ex)
        {
            Trace.WriteLine(string.Format(CultureInfo.InvariantCulture, "FontManager fallback: failed to compose font family {0}. {1}", sourceList, ex.Message));
            return fallback;
        }
    }

    private static bool HasBundledFontResources(string normalizedDirectory)
    {
        string location = "./" + normalizedDirectory;
        try
        {
            return Fonts.GetFontFamilies(ApplicationFontBaseUri, location).Any();
        }
        catch (Exception ex)
        {
            Trace.WriteLine(string.Format(CultureInfo.InvariantCulture, "FontManager fallback: failed to enumerate bundled fonts under {0}. {1}", normalizedDirectory, ex.Message));
            return false;
        }
    }

    private static string NormalizeFontDirectory(string relativeDirectory)
    {
        string normalizedDirectory = relativeDirectory.Replace('\\', '/').Trim('/');
        if (!normalizedDirectory.EndsWith("/", StringComparison.Ordinal))
        {
            normalizedDirectory += "/";
        }
        return normalizedDirectory;
    }

    private static bool ShouldWriteDiagnostics()
    {
        return string.Equals(Environment.GetEnvironmentVariable("CLIKSYNC_FONT_DIAGNOSTICS"), "1", StringComparison.Ordinal);
    }

    private static void WriteDiagnostics(FontFamily displaySans, FontFamily listDisplay, FontFamily body, FontFamily meta, FontFamily microUi)
    {
        try
        {
            List<string> diagnosticsLines = new List<string>
            {
                "Timestamp=" + DateTimeOffset.Now.ToString("O", CultureInfo.InvariantCulture),
                "Culture=" + LocalizationManager.Instance.CurrentCulture.Name
            };
            AppendDiagnostics(diagnosticsLines, "Font.DisplaySans", displaySans);
            AppendDiagnostics(diagnosticsLines, "Font.ListDisplay", listDisplay);
            AppendDiagnostics(diagnosticsLines, "Font.Body", body);
            AppendDiagnostics(diagnosticsLines, "Font.Meta", meta);
            AppendDiagnostics(diagnosticsLines, "Font.MicroUi", microUi);
            File.WriteAllLines(Path.Combine(AppContext.BaseDirectory, "font-diagnostics.txt"), diagnosticsLines);
        }
        catch (Exception ex)
        {
            Trace.WriteLine("FontManager diagnostics failed: " + ex.Message);
        }
    }

    private static void AppendDiagnostics(ICollection<string> lines, string key, FontFamily family)
    {
        if (lines == null)
        {
            return;
        }
        if (family == null)
        {
            lines.Add(key + ".Source=<null>");
            return;
        }
        lines.Add(key + ".Source=" + family.Source);
        Typeface typeface = new Typeface(family, FontStyles.Normal, FontWeights.Normal, FontStretches.Normal);
        GlyphTypeface glyphTypeface = null;
        if (typeface.TryGetGlyphTypeface(out glyphTypeface) && glyphTypeface != null)
        {
            string englishFamilyName = glyphTypeface.FamilyNames.FirstOrDefault((KeyValuePair<CultureInfo, string> entry) => string.Equals(entry.Key.IetfLanguageTag, "en-US", StringComparison.Ordinal)).Value;
            if (string.IsNullOrWhiteSpace(englishFamilyName))
            {
                englishFamilyName = glyphTypeface.FamilyNames.Values.FirstOrDefault();
            }
            lines.Add(key + ".GlyphFamily=" + englishFamilyName);
        }
        else
        {
            lines.Add(key + ".GlyphFamily=<unresolved>");
        }
    }
}





